src/Microsoft.Identity.Client/Azure/ServicePrincipal.cs

```diff
@@ -0,0 +1,248 @@
+
+            return false;
+        }
+    }
+
+    /// <summary>
+    /// IManagedIdentityConfiguration provides the configurable properties for the ManagedIdentityProbe
+    /// </summary>
+    public interface IServicePrincipalConfiguration
+    {
+        /// <summary>
+        /// CertificateBase64 is the base64 encoded representation of an x509 certificate
+        /// </summary>
+        string CertificateBase64 { get; }
+
+        /// <summary>
+        /// CertificateThumbprint is the thumbprint of the certificate
+        /// </summary>
+        string CertificateThumbprint { get; }
+
+        /// <summary>
+        /// CertificateStoreName is the name of the certificate store on Windows where the certificate is stored
+        /// </summary>
+        string CertificateStoreName { get; }
+
+        /// <summary>
+        /// CertificateStoreLocation is the location of the certificate store on Windows where the certificate is stored
+        /// </summary>
+        string CertificateStoreLocation { get; }
+
+        /// <summary>
+        /// TenantId is the AAD TenantID
+        /// </summary>
+        string TenantId { get; }
+
+        /// <summary>
+        /// ClientId is the service principal (application) ID
+        /// </summary>
+        string ClientId { get; }
+
+        /// <summary>
+        /// ClientSecret is the service principal (application) string secret
+        /// </summary>
+        string ClientSecret { get; }
+
+        /// <summary>
+        /// Authority is the URI pointing to the AAD endpoint
+        /// </summary>
+        string Authority { get; }
+    }
+
+    internal class DefaultServicePrincipalConfiguration : IServicePrincipalConfiguration
+    {
+        public string ClientId => Env.ClientId;
+
+        public string CertificateBase64 => Env.CertificateBase64;
+
+        public string CertificateThumbprint => Env.CertificateThumbprint;
+
+        public string CertificateStoreName => Env.CertificateStoreName;
+
+        public string TenantId => Env.TenantId;
+
+        public string ClientSecret => Env.ClientSecret;
+
+        public string CertificateStoreLocation => Env.CertificateStoreLocation;
+
+        public string Authority => string.IsNullOrWhiteSpace(Env.AadAuthority) ? AadAuthority.DefaultTrustedHost : Env.AadAuthority;
+    }
+
+    /// <summary>
+    /// ServicePrincipalTokenProvider fetches an AAD token provided Service Principal credentials.
+    /// </summary>
+    public class ServicePrincipalTokenProvider : ITokenProvider
+    {
+        private readonly ConfidentialClientApplication _client;
+
+        internal ServicePrincipalTokenProvider(string authority, string tenantId, string clientId, ClientCredential cred, IHttpManager httpManager)
+        {
+            _client = ConfidentialClientApplicationBuilder.Create(clientId)
+                .AddKnownAuthority(new Uri(authority), true)
+                .WithClientCredential(cred)
```

Comment:
I would not necessarily use the internal .With methods?
=> .WithCertificate and .WithClientSecret are public?

Reply:
This is only used for an internal constructor to make testing easier. The public constructors use the public .With methods.